### ISSUE TITLE
BinaryBoxProtocol.makeConnection calls startReceivingBoxes too early

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3477]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3477|
|Type|enhancement|
|Created|2008-10-16 15:12:19Z|
|Branch|https://github.com/twisted/twisted/tree/early-sendbox-3477|

The idea behind the design of `IBoxReceiver` is that you can start sending boxes as soon as you've received notification that you will be receiving boxes.

Unfortunately the actual implementation makes this impossible; if you call `sendBox` (or anything which calls it, such as `callRemote`) in `startReceivingBoxes`, `BinaryBoxProtocol` will invoke your `startReceivingBoxes` before the `transport` attribute has been set and you will get an exception.

<details><summary>Searchable metadata</summary>

```
trac-id__3477 3477
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_early_sendbox_3477 branches/early-sendbox-3477
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1224169939000000 1224169939000000
changetime__1225557427000000 1225557427000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25232]) Branching to 'early-sendbox-3477'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Fixed - barely.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

It took me a little while to understand what you meant by 'barely', and then more to understand why it wasn't trivially fixable to use an upcall to `makeConnection` in a different place.

I think this is ready to be merged.  However, I also think that it highlights a systematic flaw in poor documentation of `makeConnection` and `connectionMade` - it really is not very clear when and if one is supposed to upcall.

I think this should spawn a documentation ticket for explaining the timing and intended uses of `makeConnection`, `connectionMade`, and `startReceivingBoxes`.  Probably something that should be assigned to me.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [25248]) Merge early-sendbox-3477

Author: exarkun
Reviewer: glyph
Fixes: [#3477](https://github.com/twisted/twisted/issues/3477)
Refs: [#3476](https://github.com/twisted/twisted/issues/3476)


Fix `BinaryBoxProtocol` so that its `IBoxReceiver` can send boxes in its
`startReceivingBoxes` callback such that they are actually written to the
transport rather than result in an exception.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
